pkg/webhook/resources/node/validator.go

```diff
@@ -71,3 +95,137 @@ func validateCordonAndMaintenanceMode(oldNode, newNode *corev1.Node, nodeList []
+	if err != nil {
+		return werror.NewBadRequest(fmt.Sprintf("Failed to retrieve cpu-manager-update-status from annotation: %v", err))
+	}
+	// only validate when update status is requested
+	if updateStatus.Status != ctlnode.CPUManagerRequestedStatus {
+		return nil
+	}
+	policy := updateStatus.Policy
+
+	// check if cpu manager policy is the same
+	if err := checkCPUManagerLabel(node, policy); err != nil {
+		return err
+	}
+	// check if there is other job that still updating cpu manager policy to the same node
+	if err := checkCurrentNodeCPUManagerJobs(node, v.jobCache); err != nil {
+		return err
+	}
+	// check if this node is master and there are other master nodes undating cpu manager policy
+	// since the policy update on master node need to restart rke2-server, we only allow one master node update policy
+	if err := checkMasterNodeJobs(node, v.nodeCache, v.jobCache); err != nil {
+		return err
+	}
+	// check if there is any vm that enable cpu pinning while cpu manager is going to be disabled
+	if err := checkCPUPinningVMIs(node, policy, v.vmiCache); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkCPUManagerLabel(node *corev1.Node, policy ctlnode.CPUManagerPolicy) error {
+	cpuManagerLabel, err := strconv.ParseBool(node.Labels[kubevirtv1.CPUManager])
+	// means CPUManager feature gate not enabled
+	if err != nil {
+		return werror.NewBadRequest("label cpumanager not found")
+	}
+	// check if the cpu manager policy is the same
+	if ((policy == ctlnode.CPUManagerStaticPolicy) && cpuManagerLabel) || ((policy == ctlnode.CPUManagerNonePolicy) && !cpuManagerLabel) {
+		return werror.NewBadRequest("same cpu manager policy")
+	}
+	return nil
+}
+
+func checkCurrentNodeCPUManagerJobs(node *corev1.Node, jobCache ctlbatchv1.JobCache) error {
+	isRunning, err := isCPUManagerJobRunningOnNodes(jobCache, []string{node.Name})
+	if err != nil {
+		return werror.NewInternalError(err.Error())
+	}
+	if isRunning {
+		return werror.NewBadRequest("there is other ongoing job updating cpu manager policy")
+	}
+	return nil
+}
+
+func checkMasterNodeJobs(node *corev1.Node, nodeCache v1.NodeCache, jobCache ctlbatchv1.JobCache) error {
+	// the node is worker, no need to do validation
+	if !ctlnode.IsManagementRole(node) {
+		return nil
+	}
+
+	nodes, err := nodeCache.List(labels.Everything())
+	if err != nil {
+		return werror.NewInternalError(err.Error())
+	}
+
+	// collect master node names except the node itself
+	masterNodeNames := []string{}
+	for _, n := range nodes {
+		if n.Name != node.Name && ctlnode.IsManagementRole(n) {
+			masterNodeNames = append(masterNodeNames, n.Name)
+		}
+	}
+
+	if len(masterNodeNames) == 0 {
+		return nil
+	}
+
+	isRunning, err := isCPUManagerJobRunningOnNodes(jobCache, masterNodeNames)
+	if err != nil {
+		return werror.NewInternalError(err.Error())
+	}
+	if isRunning {
+		return werror.NewBadRequest("there is other master node updating cpu manager policy")
+	}
```

Comment:
nit: we should probably return the name of node to make it the error a bit more descriptive and ease subsequent debugging if needed.